consensus/istanbul/backend/announce.go

```diff
@@ -589,130 +649,311 @@ func (sb *Backend) regossipAnnounce(msg *istanbul.Message, payload []byte) error
-		logger.Error("Error in decoding announce versions array", "err", err)
+	err := rlp.DecodeBytes(payload, &signedAnnVersions)
+	if err != nil {
+		logger.Warn("Error in decoding received Signed Announce Versions msg", "err", err)
+		return err
 	}
 
-	logger.Trace("Handling an AnnounceVersions message", "announceVersions", fmt.Sprintf("%v", announceVersions))
-
 	// If the announce's valAddress is not within the validator connection set, then ignore it
 	validatorConnSet, err := sb.retrieveValidatorConnSet()
 	if err != nil {
+		logger.Trace("Error in retrieving validator conn set", "err", err)
 		return err
 	}
 
-	announcesToRequest := make([]common.Address, 0)
-	for _, announceVersion := range announceVersions {
-		// Ignore this announceVersion entry if it's val address is not in the current validator connection set.
-		if !validatorConnSet[announceVersion.ValAddress] {
-			logger.Trace("Ignoring announceVersion since it's not in active/elected valset", "valAddress", announceVersion.ValAddress)
+	var validEntries []*vet.SignedAnnounceVersionEntry
+	validAddresses := make(map[common.Address]bool)
+	// Verify all entries are valid and remove duplicates
+	for _, signedAnnVersion := range signedAnnVersions {
+		err := signedAnnVersion.ValidateSignature()
+		if err != nil {
+			logger.Debug("Error validating signed announce version signature", "address", signedAnnVersion.Address, "err", err)
+			continue
+		}
+		if !validatorConnSet[signedAnnVersion.Address] {
+			logger.Debug("Found signed announce version from an address not in the validator conn set", "address", signedAnnVersion.Address)
+			continue
+		}
+		if _, ok := validAddresses[signedAnnVersion.Address]; ok {
+			logger.Debug("Found duplicate signed announce version in message", "address", signedAnnVersion.Address)
 			continue
 		}
+		validAddresses[signedAnnVersion.Address] = true
+		validEntries = append(validEntries, signedAnnVersion.Entry())
+	}
+	return sb.upsertSignedAnnounceVersionEntries(validEntries)
+}
 
-		sb.cachedAnnounceMsgsMu.RLock()
-		if cachedEntry, ok := sb.cachedAnnounceMsgs[announceVersion.ValAddress]; !ok || cachedEntry.MsgVersion < announceVersion.AnnounceMsgVersion {
-			announcesToRequest = append(announcesToRequest, announceVersion.ValAddress)
+func (sb *Backend) upsertSignedAnnounceVersionEntries(entries []*vet.SignedAnnounceVersionEntry) error {
+	logger := sb.logger.New("func", "upsertSignedAnnounceVersions")
+	newEntries, err := sb.signedAnnounceVersionTable.Upsert(entries)
+	if err != nil {
+		logger.Warn("Error in upserting entries", "err", err)
+	}
+
+	// If this node is a validator (checked later as a result of this call) and it receives a signed announce
+	// version from a remote validator that is newer than the remote validator's
+	// version in the val enode table, this node did not receive a direct announce
+	// and needs to announce its own enode to the remote validator.
+	sb.startGossipAnnounceTask()
```

Comment:
Given the 5 minute DDoS cooldown, and the fact that this announcement may contain many more enodes than just the ones that changed, I dont think we should trigger this here.

Reply:
I think it'd be nice to trigger an announce task the moment we become aware of an inconsistency between the signed announce version table and the validator enode table. Where else would you suggest? Will think about this more